### ISSUE TITLE
Convert JavaScript to Kotlin

### ISSUE THREAD

**Issue by BFergerson** [sourceplusplus/SourceMarker-Alpha#12]
### Objective

Three pages need to be converted to Kotlinx.html, which are:
- [ ] overview_view.js
- [ ] traces_view.js
- [ ] configuration_view.js
- [ ] overview.js
- [ ] traces.js
- [ ] configuration.js
- [ ] source_eventbus_bridge.js

### Information
Save source_eventbus_bridge.js for last as it warrants further thinking and discussion.

**Comment by BFergerson** [comment]
@chess-equality, you should have write access to this repo. Create a new branch or fork to work on this issue.

I've moved MockMonitorServer to the test directory and I've changed the portal_uuid to be null. Now you can access the portal overview by just going to `http://localhost:8080`.

Let me know once you have everything setup.

btw, when you do PRs make sure you also comment that it is ready and not just mark as needs review. I missed your last request so I didn't know you were finished.

**Comment by BFergerson** [comment]
Also, I've left hints at what types certain variables should have once you've converted to Kotlin. Example:
https://github.com/sourceplusplus/SourceMarker-Alpha/blob/c31fc7a8a2948467f2c05935db52a8fbd28f4d6d/portal/src/main/resources/webroot/js/views/overview_view.js#L109

That `card` variable should be:
https://github.com/sourceplusplus/SourceMarker-Alpha/blob/c31fc7a8a2948467f2c05935db52a8fbd28f4d6d/protocol/src/commonMain/kotlin/com/sourceplusplus/protocol/portal/BarTrendCard.kt#L9

**Comment by BFergerson** [comment]
And another thing, I created temporary `author: me` to all of code. Feel free to update to your name if you wish.

**Comment by chess-equality** [comment]
@BFergerson I have forked and set it up with IDEA. I have some questions:

1. How to run the portal? Run `main` on `PortalServer`?

By the way, I get the error `Unresolved reference: readAllBytes` on this line:
https://github.com/sourceplusplus/SourceMarker-Alpha/blob/master/portal/src/main/kotlin/com/sourceplusplus/portal/backend/PortalServer.kt#L67

2. What would be the approach?

> Once the code is copied over verbatim

I have found this:
https://superkotlin.com/kotlin-javascript/

It also talks about passing a data class to a JS function.

Also https://github.com/kotlin/kotlinx.html/wiki/Style-and-script-tags

**Comment by BFergerson** [comment]
You can run the portal by running 'PortalServer' directly or using the Gradle command `gradle portal:run`. You can avoid building the whole project by building the portal directly via `gradle portal:build`.

You'll need to use JDK.11+ to fix `readAllBytes` issue.

By copying over the code verbatim I mean a direct translation with no optimizations/cleaning/merging/etc. Verbatim translation.

**Comment by BFergerson** [comment]
@chess-equality, actually looks like it might be JDK.13+

**Comment by chess-equality** [comment]
@BFergerson I have run it with GraalVM JDK 11, ok will use JDK 13.

**Comment by BFergerson** [comment]
@chess-equality, it's actually building on JDK 11 https://github.com/sourceplusplus/SourceMarker-Alpha/blob/71365ca82cd9a3768b503ae43f965c7525081a5f/.github/workflows/build.yml#L39.

I think there is just something weird with that `readAllBytes` method. I'll probably end up removing it.

**Comment by BFergerson** [comment]
@chess-equality, comment out the static handler I added and uncomment the static handler you added. See if that gets rid of the issue. I had to add that static handler to make it work correctly in the plugin but since you don't need the plugin you can work with the old static handler.

**Comment by BFergerson** [comment]
@chess-equality, so now it works with JDK 11 using `readAllBytes`? What was the issue? And if it works as is just go with that. If it doesn't then you know a temporary fix.

**Comment by chess-equality** [comment]
@BFergerson I was using JDK 8, now I have switched to JDK 11.

**Comment by BFergerson** [comment]
@chess-equality, notice you're writing code like:
```
script(type = ScriptType.textJavascript) {
...
```

Please understand that this ticket does not involve moving JavaScript code into Kotlin.
This ticket involves translating JavaScript code to Kotlin.

Example:
```javascript
function clickedViewAverageThroughputChart() {
    console.log("Clicked view average throughput");
    currentMetricType = "Throughput_Average";
    eb.send('SetActiveChartMetric', {'portal_uuid': portalUuid, 'metric_type': currentMetricType});
}
```
To:
```kotlin
fun clickedViewAverageThroughputChart() {
    console.log("Clicked view average throughput");
    currentMetricType = "Throughput_Average";
    //vertx.eventBus().send('SetActiveChartMetric', {'portal_uuid': portalUuid, 'metric_type': currentMetricType});
}
```

I created the frontend module to make it easier to get started on this issue. I added an example method. These:
https://github.com/sourceplusplus/SourceMarker-Alpha/blob/9700ef95fbedc808636f9c1d79d3ec0997250d75/portal/backend/src/main/resources/webroot/js/overview.js#L3-L4
https://github.com/sourceplusplus/SourceMarker-Alpha/blob/9700ef95fbedc808636f9c1d79d3ec0997250d75/portal/backend/src/main/resources/webroot/js/overview.js#L49-L53

After translation turned into this: https://github.com/sourceplusplus/SourceMarker-Alpha/blob/master/portal/frontend/src/main/kotlin/frontend/display/views/OverviewView.kt

Something like this:
```javascript
$(document).ready(function () {
    $('.ui.calendar').calendar();
    $('#last_5_minutes_time').removeClass('active');
}
```

Would translate to:
```kotlin
jq().ready {
    jq(".ui.calendar").calendar()
    jq("#last_5_minutes_time").removeClass("active")
}
```

Notice there is no use of scripttype/unsafe/etc. The goal is to convert/translate JavaScript to pure Kotlin.

Here is a link on interacting with the DOM in Kotlin: https://kotlinlang.org/docs/tutorials/javascript/browser-api-dom.html

**Comment by BFergerson** [comment]
You're going to have to comment on Vert.x code for now. I'm going to be working on getting that to work. In the meantime just copy it over and comment it out.

**Comment by chess-equality** [comment]
@BFergerson 
> notice you're writing code like:
Yes I realized it's not quite like it, so I have stopped tracking my time on that.

Thanks for this approach.

**Comment by BFergerson** [comment]
@chess-equality, added more example code:
https://github.com/sourceplusplus/SourceMarker-Alpha/blob/51dbb82571fbfc23685cac40b09031fdbc9d1289/portal/frontend/src/main/kotlin/ExampleJS.kt#L4-L19

Which does this:
![Peek 2020-09-24 10-26](https://user-images.githubusercontent.com/3278877/94108498-79a64500-fe50-11ea-8abf-69badee5e061.gif)

**Comment by BFergerson** [comment]
Basic Vert.x functionality now working in Kotlin JS:
https://github.com/sourceplusplus/SourceMarker-Alpha/blob/36fb1f991edaf34faba002c98b38842eb68d5eb3/portal/frontend/src/main/kotlin/frontend/display/OverviewDisplay.kt#L9-L44

Which is a partial translation of:
https://github.com/sourceplusplus/SourceMarker-Alpha/blob/36fb1f991edaf34faba002c98b38842eb68d5eb3/portal/backend/src/main/resources/webroot/js/overview.js#L6-32

---

Note: My usage of `js("")` is a crutch because those functions haven't been translated to Kotlin yet so I'm just calling them in JavaScript so the UI still works. Everything in `js("")` needs to be translated to Kotlin so you can call it like a regular Kotlin function.

**Comment by BFergerson** [comment]
@chess-equality, I've converted `overview.js` to `OverviewPage.kt`. That should be enough to see how to accomplish this task.

**Comment by chess-equality** [comment]
@BFergerson How do I run `SourcePortalUI`? `main` is not runnable from the IDE.

Also could you please describe a little bit the structure, i.e., `frontend`(`SourcePortalUI`) and `backend`(`PortalServer`).

Thanks.

**Comment by BFergerson** [comment]
@chess-equality, check chat. I've been trying to reach you.

**Comment by BFergerson** [comment]
@chess-equality, I've updated the repository one last time. This will be the final time I do this. Please re-fork and copy/paste your changes back onto this repository. You shouldn't have any merging issues.
